compiler/rustc_builtin_macros/src/env.rs

```diff
@@ -63,10 +63,19 @@ pub fn expand_env<'cx>(
         Some(exprs) => exprs.into_iter(),
     };
 
-    let Some((var, _style)) = expr_to_string(cx, exprs.next().unwrap(), "expected string literal") else {
+    let var_expr = exprs.next().unwrap();
+    let Some((mut var, _style)) = expr_to_string(cx, var_expr.clone(), "expected string literal") else {
         return DummyResult::any(sp);
     };
 
+    let ast::ExprKind::Lit(ast::token::Lit { kind: ast::token::LitKind::Str, symbol: original_var, ..}) = &var_expr.kind else {
+        unreachable!("`expr_to_string` ensures this is a string lit")
+    };
+    if original_var.as_str().contains('\\') {
+        // If the variable name contains escape sequences, we should report it to the user
+        // as written in the code to avoid confusing diagnostics
+        var = *original_var;
+    }
```

Comment:
Why don't we always prefer reporting using original_var?

Reply:
Hmm, that sounds more sensible